src/QsCompiler/CompilerOptimizations/PureCircuitAPI.fs

```diff
@@ -0,0 +1,211 @@
+        | t, Range | t, ArrayType _ -> t
+        | ArrayType t, Int -> t.Resolution
+        | _ -> failwithf "Invalid ArrayItem node"
+    | QubitReference (Qubit _) -> TypeKind.Qubit
+    | QubitReference (QubitArray _) -> ArrayType (ResolvedType.New TypeKind.Qubit)
+    constToTypeKind c |> ResolvedType.New
+
+
+/// Returns the Q# expression corresponding to the given Const
+let rec private constToExpr (cc: CircuitContext) (c: Const): TypedExpression =
+    let constToExprKind = function
+    | UnitValue -> Expr.UnitValue
+    | ValueTuple cl -> cl |> Seq.map (constToExpr cc) |> ImmutableArray.CreateRange |> Expr.ValueTuple
+    | IntLiteral x -> Expr.IntLiteral x
+    | BigIntLiteral x -> Expr.BigIntLiteral x
+    | DoubleLiteral x -> Expr.DoubleLiteral x
+    | BoolLiteral x -> Expr.BoolLiteral x
+    | StringLiteral x -> Expr.StringLiteral (NonNullable<_>.New x, ImmutableArray.Empty)
+    | ResultLiteral x -> Expr.ResultLiteral x
+    | PauliLiteral x -> Expr.PauliLiteral x
+    | RangeLiteral (x, y) -> Expr.RangeLiteral (constToExpr cc x, constToExpr cc y)
+    | ValueArray cl -> cl |> Seq.map (constToExpr cc) |> ImmutableArray.CreateRange |> Expr.ValueArray
+    | ArrayItem (x, y) -> Expr.ArrayItem (constToExpr cc x, constToExpr cc y)
+    | QubitReference (Qubit i) -> Identifier (LocalVariable (NonNullable<_>.New cc.qubitReferences.[i]), Null)
+    | QubitReference (QubitArray i) -> Identifier (LocalVariable (NonNullable<_>.New cc.qubitReferences.[i]), Null)
+    wrapExpr (constToType c).Resolution (constToExprKind c)
+
+
+/// Returns the Q# expression correponding to the given gate call
+let private gateCallToExpr (cc: CircuitContext) (gc: GateCall): TypedExpression =
+    let mutable method = wrapExpr UnitType (Identifier (GlobalCallable gc.gate, Null))
+    let mutable arg = constToExpr cc gc.arg
+
+    if gc.adjoint then
+        method <- wrapExpr UnitType (AdjointApplication method)
+    for control in gc.controls do
+        method <- wrapExpr UnitType (ControlledApplication method)
+        arg <- constToExpr cc (ValueTuple [control; exprToConst (ref cc) arg |> Option.get])
+
+    wrapExpr UnitType (CallLikeExpression (method, arg))
+
+
+/// Returns the list of Q# expressions corresponding to the given circuit
+let private circuitToExprList (cc: CircuitContext) (circuit: Circuit): TypedExpression list =
+    List.map (gateCallToExpr cc) circuit.gates
+
+
+/// Given a pure circuit, performs basic optimizations and returns the new circuit
+let private optimizeCircuit (circuit: Circuit): Circuit =
+    let mutable circuit = circuit
+    let mutable i = 0
+    while i < circuit.gates.Length - 1 do
+        if circuit.gates.[i] = { circuit.gates.[i+1] with adjoint = not circuit.gates.[i+1].adjoint } then
+            circuit <- { circuit with gates = removeIndices [i; i+1] circuit.gates}
+        elif circuit.gates.[i] = circuit.gates.[i+1] && List.contains circuit.gates.[i].gate.Name.Value ["X"; "Z"; "H"; "CNOT"] then
```

Comment:
Ideally you could leverage the operation metadata tracked in the Q# compilation here rather than hardcode this list.

Reply:
Added to todo list